frontend_tests/zjsunit/namespace.js

```diff
@@ -79,6 +80,14 @@ function load(request, parent, isMain) {
         return obj;
     }
 
+    if (filename.endsWith(".png") || filename.endsWith(".svg")) {
+        // Module._resolveFilename resolves symlinks, which isn't what
+        // we want here -- reconstruct the resolved path as simply as
+        // possible.
+        const paths = Module._resolveLookupPaths(request, parent);
+        const resolved_filename = path.resolve(paths[0], request);
+        return resolved_filename.replace(root_dir + "/static/", "/static/");
+    }
```

Comment:
Sure.  But I thought the goal for zjsunit here was to create _a_ believable URL for the PNG or SVG file.  Without this change:
```
running test example4
running test compose_ui
running test dispatch
running test copy_and_paste
/srv/zulip-emoji-cache/6dea1e5fc556471d6d40b3b3e4e3dd52e72323b4/emoji/images-google-64/zulip.png:1
�PNG


SyntaxError: Invalid or unexpected token
    at Object.compileFunction (node:vm:360:18)
    at wrapSafe (node:internal/modules/cjs/loader:1088:15)
    at Module._compile (node:internal/modules/cjs/loader:1123:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Object.newLoader [as .js] (/srv/zulip-npm-cache/51aaca0fd2a5e5e1e2e7d44f654c60a5dfb3d1d0/node_modules/pirates/lib/index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Function.load [as _load] (/home/zulipdev/zulip/frontend_tests/zjsunit/namespace.js:92:12)
    at Module.require (node:internal/modules/cjs/loader:1061:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at _zulip (/home/zulipdev/zulip/static/js/emoji.js:3:1)
    at _get_original__ (/home/zulipdev/zulip/static/js/emoji.js:320:2)
    at _get__ (/home/zulipdev/zulip/static/js/emoji.js:320:2)
    at Object.<anonymous> (/home/zulipdev/zulip/static/js/emoji.js:77:14)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module._compile (/srv/zulip-npm-cache/51aaca0fd2a5e5e1e2e7d44f654c60a5dfb3d1d0/node_modules/pirates/lib/index.js:136:24)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Object.newLoader [as .js] (/srv/zulip-npm-cache/51aaca0fd2a5e5e1e2e7d44f654c60a5dfb3d1d0/node_modules/pirates/lib/index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Function.load [as _load] (/home/zulipdev/zulip/frontend_tests/zjsunit/namespace.js:92:12)
    at Module.require (node:internal/modules/cjs/loader:1061:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at exports.zrequire (/home/zulipdev/zulip/frontend_tests/zjsunit/namespace.js:271:12)
    at Object.<anonymous> (/home/zulipdev/zulip/frontend_tests/node_tests/dispatch.js:107:15)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module._compile (/srv/zulip-npm-cache/51aaca0fd2a5e5e1e2e7d44f654c60a5dfb3d1d0/node_modules/pirates/lib/index.js:136:24)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Object.newLoader [as .js] (/srv/zulip-npm-cache/51aaca0fd2a5e5e1e2e7d44f654c60a5dfb3d1d0/node_modules/pirates/lib/index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Function.load [as _load] (/home/zulipdev/zulip/frontend_tests/zjsunit/namespace.js:92:12)
    at Module.require (node:internal/modules/cjs/loader:1061:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at run_one_module (/home/zulipdev/zulip/frontend_tests/zjsunit/index.js:98:5)
(...)
```

My assumption was that we needed to make it a potential path for this mock, not cover every case.

Reply:
To be concrete, my suggestion is to replace the mock with something much simpler, rather than spending extra effort to maintain an invariant that webpack itself does not maintain.

```diff
--- a/frontend_tests/node_tests/emoji.js
+++ b/frontend_tests/node_tests/emoji.js
@@ -78,3 +78,3 @@ run_test("get_emoji_details_by_name", () => {
         emoji_code: "zulip",
-        url: "/static/generated/emoji/images/emoji/unicode/zulip.png",
+        url: require.resolve("../../static/generated/emoji/images/emoji/unicode/zulip.png"),
     });
--- a/frontend_tests/node_tests/reactions.js
+++ b/frontend_tests/node_tests/reactions.js
@@ -360,3 +360,3 @@ test("sending", ({override, override_rewire}) => {
             emoji_code: "zulip",
-            url: "/static/generated/emoji/images/emoji/unicode/zulip.png",
+            url: require.resolve("../../static/generated/emoji/images/emoji/unicode/zulip.png"),
         });
@@ -864,3 +864,3 @@ test("view.insert_new_reaction (them w/zulip emoji)", ({mock_template}) => {
             count: 1,
-            url: "/static/generated/emoji/images/emoji/unicode/zulip.png",
+            url: require.resolve("../../static/generated/emoji/images/emoji/unicode/zulip.png"),
             is_realm_emoji: true,
--- a/frontend_tests/zjsunit/namespace.js
+++ b/frontend_tests/zjsunit/namespace.js
@@ -83,8 +83,3 @@ function load(request, parent, isMain) {
     if (filename.endsWith(".png") || filename.endsWith(".svg")) {
-        // Module._resolveFilename resolves symlinks, which isn't what
-        // we want here -- reconstruct the resolved path as simply as
-        // possible.
-        const paths = Module._resolveLookupPaths(request, parent);
-        const resolved_filename = path.resolve(paths[0], request);
-        return resolved_filename.replace(root_dir + "/static/", "/static/");
+        return filename;
     }
```

Actually trying to do this reveals a bug: the frontend Markdown processor is now inconsistent with the backend.

```console
$ tools/test-js-with-node markdown
Starting node tests...
running test markdown
        test: markdown_detection
        test: marked_shared
--------------------------------------------------
test failed: /srv/zulip/frontend_tests/node_tests/markdown.js > marked_shared

AssertionError [ERR_ASSERTION]: Failure in test: zulip_emoji
Actual and expected output do not match.  Showing diff
- <p><img alt=":zulip:" class="emoji" src="/srv/zulip-emoji-cache/6dea1e5fc556471d6d40b3b3e4e3dd52e72323b4/emoji/images-google-64/zulip.png" title="zulip"></p>
+ <p><img alt=":zulip:" class="emoji" src="/static/generated/emoji/images/emoji/unicode/zulip.png" title="zulip"></p>
    at new AssertionError (node:internal/assert/assertion_error:452:5)
…
```

So I don’t think we want to webpackify `emoji/unicode/zulip.png` at all, at least while the backend Markdown processor is still in the business of resolving emoji URLs.

Reply:
Hm, good point.  I've reverted that hunk, and left a comment there about why we're leaving it as-is.